posthog/async_migrations/migrations/0002_fill_pdi2_from_pdi.py

```diff
@@ -0,0 +1,140 @@
+        2019-01-01 10:10:10   |  A           |  1          |  1
+
+        Hence there is no way of telling if the person was first deleted then
+        re-associated or the other way around. Hence at the moment with the old
+        table we're just checking that all rows relating to a (distinct_id,
+        person_id) pair are marked as is_deleted = 0.
+
+        The person_distinct_id table uses the `CollapsingMergeTree` engine, which
+        will at some point pair these two rows together, but we can't rely on
+        waiting. It also means that if we've have to be diligent with ensuring
+        we have written cancel rows correctly. It's good for cases where we
+        might want to perform aggregation over data, but `ReplacingMergeTree` is
+        a better fix where we just want the latest data.
+
+        The new schema includes a `version` column which should be strictly
+        increasing. At the time of writing, this version is a big int 
+        [updated in a
+        transaction](https://github.com/PostHog/posthog/blob/bcf2b6370f8d2205f1f7d5fb5f431124c3848691/plugin-server/src/worker/ingestion/properties-updater.ts#L58:L58)
+        within postgres, then propagated to clickhouse on successful commit via
+        kafka.
+
+        NOTE: on cloud we have already migrated team_id = 2, but adding
+        additional rows should not be an issue as we're adding these with
+        version = 0.
+
+        The migration strategy:
+
+         1. write to both pdi and pdi2 any new updates (already done separate to
+            this migration)
+         2. read all non-deleted (team_id, distinct_id, person_id) rows from pdi
+            into pdi2 (this migration)
+         3. update app to always use pdi2 (separate PR)
+    """
+
+    posthog_min_version = "1.31.0"
+    posthog_max_version = "1.32.0"
+
+    service_version_requirements = [
+        ServiceVersionRequirement(service="clickhouse", supported_version=">=21.6.0"),
+    ]
+
+    operations = [
+        AsyncMigrationOperation(
+            database=AnalyticsDBMS.CLICKHOUSE,
+            sql=f"""
+                INSERT INTO person_distinct_id2(
+                    team_id, 
+                    distinct_id, 
+                    person_id,
+                    is_deleted, 
+                    version
+                )
+                
+                SELECT
+                    any(team_id) as team_id,
+                    distinct_id,
+                    argMax(person_id, _timestamp) as person_id,
+                    0 as is_deleted,
+                    0 as version
+
+                FROM
+                    (
+                        /*
+                            Get all team_id, distinct_id, person_id tuples that 
+                            haven't been deleted.
+                            
+                            Also, use the latest timestamp.
+                            
+                            Note that this query differs from the existing 
```

Comment:
Q: Why does it differ? I'm not confident this is actually correct. The scenario I'm considering is that:
1. Person gets merged
2. We emit rows with equal timestamps.

Note that there have been man-months spent on proving and making the previous query correct by Yakko/Tiina.

I posit it's easier and safer to just do the same thing we do in queries and not reinvent the wheel - reducing risk here is important.

Reply:
This is the query from https://github.com/PostHog/posthog/issues/7663

The comment is in reference to being different to https://github.com/PostHog/posthog/blob/78a787bc3bb2d6eaefb8efd2114453e21e58ac33/ee/clickhouse/sql/person.py#L255

E.g. it does not have the further wrapping. If I add a comment as to the source of this and explicitly the difference would this be enough?

Reply:
I've added a comment regarding oom considerations

Reply:
Differences to https://github.com/PostHog/posthog/issues/7663:

 1. doesn't filter by team_id
 2. groups by team_id

Reply:
Ah! Sorry for getting this wrong!